ecs-agent/tcs/handler/handler.go

```diff
@@ -150,39 +172,97 @@ func (session *telemetrySession) StartTelemetrySession(ctx context.Context, endp
+	})
+
+	// newDisconnectTimeoutTimerHandler returns a timer.Afterfunc(timeout, f) which will
+	// call f as goroutine after timeout. The timeout is currently set to 30m+jitter(5m max) to match max duration
+	// of connection with TCS. This timer is meant to handle s.startTCSSession running in blocking mode
+	// beyond 30m+jitter as s.startTCSSession has 2 possible paths: returns with error or continue running
+	// in blocking mode.
+	// Happy path: it returns with error, then timer stops, goroutine to disconnect never starts.
+	// Edge case: it continues to run beyond the maximum duration of TCS connection. Timer starts goroutine
+	// from DisconnectTimeoutTimer to disconnect; guard against hanging connection to unhealthy TCS host.
+	disconnectTimer := session.newDisconnectTimeoutHandler(client, startTime)
+	defer disconnectTimer.Stop()
+
 	// start a timer and listens for tcs heartbeats/acks. The timer is reset when
 	// we receive a heartbeat from the server or when a published metrics message
 	// is acked.
-	timer := time.NewTimer(retry.AddJitter(session.heartbeatTimeout, session.heartbeatJitterMax))
-	defer timer.Stop()
-	client.AddRequestHandler(heartbeatHandler(timer, session.heartbeatTimeout, session.heartbeatJitterMax))
-	client.AddRequestHandler(ackPublishMetricHandler(timer, session.heartbeatTimeout, session.heartbeatJitterMax))
-	client.AddRequestHandler(ackPublishHealthMetricHandler(timer, session.heartbeatTimeout, session.heartbeatJitterMax))
-	client.AddRequestHandler(ackPublishInstanceStatusHandler(timer, session.heartbeatTimeout, session.heartbeatJitterMax))
+	heartBeatTimer := session.newHeartbeatTimeoutHandler(client, startTime)
+	defer heartBeatTimer.Stop()
+
+	client.AddRequestHandler(heartbeatHandler(heartBeatTimer, session.heartbeatTimeout, session.heartbeatJitterMax))
+	client.AddRequestHandler(ackPublishMetricHandler(heartBeatTimer, session.heartbeatTimeout, session.heartbeatJitterMax))
+	client.AddRequestHandler(ackPublishHealthMetricHandler(heartBeatTimer, session.heartbeatTimeout, session.heartbeatJitterMax))
+	client.AddRequestHandler(ackPublishInstanceStatusHandler(heartBeatTimer, session.heartbeatTimeout, session.heartbeatJitterMax))
 	client.SetAnyRequestHandler(anyMessageHandler(client, wsRWTimeout))
-	serveC := make(chan error, 1)
-	go func() {
-		serveC <- client.Serve(ctx)
-	}()
-	select {
-	case <-ctx.Done():
-		// outer context done, agent is exiting
-		client.Disconnect()
-	case <-timer.C:
-		seelog.Info("TCS Connection hasn't had any activity for too long; disconnecting")
-		client.Disconnect()
-	case err := <-serveC:
-		return err
+	return client.Serve(ctx)
+}
+
+func (session *telemetrySession) getTelemetryEndpoint() (string, error) {
+	containerInstanceARN := session.containerInstanceArn
+	tcsEndpoint, err := session.ecsClient.DiscoverTelemetryEndpoint(containerInstanceARN)
+	if err != nil {
+		logger.Error("tcs: unable to discover poll endpoint", logger.Fields{
```

Comment:
Will we log this error in other callers?

Reply:
Sorry, did not get you completely. Which other callers? There is no other non-test function where we invoke DiscoverTelemetryEndpoint().

Reply:
Sry I mean will the caller of the `getTelemetryEndpoint` print the error as well? Just want to confirm we won't log the same error twice. But it's not a blocker. Probably we can also wrap this error with new message.

Reply:
Got it. It's not being logged in the caller, but it is being logged in the caller of that function, i.e. Start(). Sure, let me pick it up in follow-up PR. Thank you :)